_posts/2022-06-13-aws-rds-iam-authentication-spring-boot.md

```diff
@@ -0,0 +1,121 @@
+On EKS (Kubernetes for AWS), you can use [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html){:target="_blank" rel="noopener noreferrer"}.
+
+
+## Spring Boot
+To achieve the token fetching in Spring Boot, I created a custom `HikariDataSource` which overrides the `getPassword()` method and used the AWS Java SDK to create a request for an authentication token from RDS.
+
+Note that for MariaDB, you do not need your own custom implementation as [they have an implementation of their own](https://mariadb.com/kb/en/mariadb-connector-j-250-release-notes/#aws-iam){:target="_blank" rel="noopener noreferrer"}. 
+
+```java
+public class RdsIAMDataSource extends HikariDataSource {
+    @Override
+    public String getPassword() {
+        RdsIamAuthTokenGenerator authTokenGenerator = RdsIamAuthTokenGenerator.builder()
+                .credentials(new DefaultAWSCredentialsProviderChain())
+                .region(new DefaultAwsRegionProviderChain().getRegion())
+                .build();
+
+        URI jdbcUri = parseJdbcURL(getJdbcUrl());
+
+        GetIamAuthTokenRequest iamAuthTokenRequest = GetIamAuthTokenRequest.builder()
+                .hostname(jdbcUri.getHost())
+                .port(jdbcUri.getPort())
+                .userName(getUsername())
+                .build();
+
+        return authTokenGenerator.getAuthToken(iamAuthTokenRequest);
+    }
+
+    private URI parseJdbcURL(String jdbcUrl) {
+        String uri = jdbcUrl.substring(5);
+        return URI.create(uri);
+    }
+}
+```
+
+To make sure that our connection does not drop due to authentication failures, we are going to set the maximum lifespan for the connection (remember that authentication tokens are only valid for 15 minutes).
+
+Don't forget to set your database properties (host, port and username). 
+**Important note:** IAM authentication to the database requires an SSL connection.
+
+```properties
+# Sets the max lifetime of each connection to 840000 ms (14 minutes). After 14 minutes, the application does a new request to RDS for a fresh authentication token
+spring.datasource.hikari.max-lifetime=840000
+# JDBC url with the properties 'ssl' and 'sslmode', this properties are bound to PostgreSQL instances
+spring.datasource.url=jdbc:postgresql://<db-instance-name>.<random>.<region>.rds.amazonaws.com:<port>/<db-name>?ssl=true&sslmode=require
+# Database user which has the 'rds_iam' role
+spring.datasource.username=<db user>
+```
+With this implementation, the `getPassword()` method will be called every 14 minutes.
+
+## Conclusion
+Security is an important factor in many companies and environments and this is definitely a recommended way of securely accessing your database from inside your application.
+While it is slightly more of a hassle to set it up, it is definitely worth your while to implement RDS IAM authentication, especially if you are already running your application and infrastructure on AWS.
```

Comment:
Maybe add that you don't have to worry about managing passwords anymore (e.g. rotating, ...)